### PR TITLE
Added Tag/Untag Command

### PR DESCRIPTION
**Pull request by krusagiz**
Added cumulative `tag` command

Added selective `untag` command

Tests are included

### CODE REVIEW

File: src/main/java/thrift/logic/commands/TagCommand.java [hunk 1]
```diff
@@ -0,0 +1,133 @@
...
+                && index.equals(((TagCommand) other).index)
+                && tagSet.equals(((TagCommand) other).tagSet));
+    }
+
+    /**
+     * Creates and returns a {@code Transaction}
+     * with the new tags from {@code tagSet} appended without duplicates.
+     */
+    private Transaction createTaggedTransaction(Transaction transactionToTag,
+                                                Set<Tag> tagSet) throws CommandException {
```
Comment (lightz96) [inline src/main/java/thrift/logic/commands/TagCommand.java]: I think the indention is wrong

File: src/test/java/thrift/testutil/TagSetBuilder.java [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+    /**
+     * Creates a blank TagSetBuilder with no tags
+     */
+    public TagSetBuilder() {
+        this.tagSet = new HashSet<Tag>();
+    }
+
+    /**
+     * Creates a TagSetBuilder with tags made from {@param tagNames}
+     * @param tagNames whose tags are used as reference
```
Comment (lightz96) [inline src/test/java/thrift/testutil/TagSetBuilder.java:31]: Do leave an empty line between description and parameter section.

File: src/main/java/thrift/logic/commands/UntagCommand.java [hunk 3]
```diff
@@ -0,0 +1,130 @@
...
+    public static final String MESSAGE_NOT_UNTAGGED = "At least one tag must be provided.";
+    public static final String MESSAGE_NO_DEL_TAGS = "Specified tags do not exist, the transaction was not updated.";
+
+    private final Index index;
+    private final Set<Tag> tagSet;
+    private final StringBuilder nonexistentTags;
+
+    /**
+     * Creates a UntagCommand to untag the specified {@code Transaction}
+     * @param index of the transaction in the filtered transaction list to update
```
Comment (lightz96) [inline src/main/java/thrift/logic/commands/UntagCommand.java:52]: Do leave an empty line between description and parameter section.

File: src/main/java/thrift/logic/parser/TagCommandParser.java [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+
+/**
+ * Parses input arguments and creates a new TagCommand object.
+ */
+public class TagCommandParser implements Parser<TagCommand> {
+
+    /**
+     * Parses the given {@code String} of arguments in the context of the TagCommand
+     * and returns an TagCommand object for execution.
+     * @throws ParseException if the user input does not conform the expected format
```
Comment (lightz96) [inline src/main/java/thrift/logic/parser/TagCommandParser.java:23]: Do leave an empty line between description and parameter section.

File: src/main/java/thrift/logic/parser/UntagCommandParser.java [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+
+/**
+ * Parses input arguments and creates a new UntagCommand object.
+ */
+public class UntagCommandParser implements Parser<UntagCommand> {
+
+    /**
+     * Parses the given {@code String} of arguments in the context of the UntagCommand
+     * and returns an UntagCommand object for execution.
+     * @throws ParseException if the user input does not conform the expected format
```
Comment (lightz96) [inline src/main/java/thrift/logic/parser/UntagCommandParser.java:23]: Do leave an empty line between description and parameter section.

File: src/test/java/thrift/testutil/TagSetBuilder.java [hunk 6]
```diff
@@ -0,0 +1,75 @@
...
+     */
+    public TagSetBuilder(String... tagNames) {
+        this.tagSet = (Stream.of(tagNames).map(Tag::new).collect(Collectors.toSet()));
+    }
+
+    /**
+     * Creates a TagSetBuilder with tags from {@param tagSet}
+     * @param tagSet whose tags are used as reference
```
Comment (lightz96) [inline src/test/java/thrift/testutil/TagSetBuilder.java:40]: Do leave an empty line between description and parameter section.

File: src/test/java/thrift/testutil/TagSetBuilder.java [hunk 7]
```diff
@@ -0,0 +1,75 @@
...
+     */
+    public TagSetBuilder(Set<Tag> tagSet) {
+        this.tagSet = tagSet;
+    }
+
+    /**
+     * Creates a TagSetBuilder with tags from {@param transaction}
+     * @param transaction whose tags are used as reference
```
Comment (lightz96) [inline src/test/java/thrift/testutil/TagSetBuilder.java:49]: Do leave an empty line between description and parameter section.

File: src/test/java/thrift/testutil/TagSetBuilder.java [hunk 8]
```diff
@@ -0,0 +1,75 @@
...
+     */
+    public TagSetBuilder(Transaction transaction) {
+        this.tagSet = new HashSet<Tag>(transaction.getTags());
+    }
+
+    /**
+     * Returns a {@code TagSetBuilder} with a tag having tagName appended
+     * @param tagName to be made into a {@code Tag} object and appended
```
Comment (lightz96) [inline src/test/java/thrift/testutil/TagSetBuilder.java:58]: Do leave an empty line between description and parameter section.

File: src/test/java/thrift/testutil/TagSetBuilder.java [hunk 9]
```diff
@@ -0,0 +1,75 @@
...
+     * @return {@code TagSetBuilder} with a tag having tagName appended
+     */
+    public TagSetBuilder addTag(String tagName) {
+        tagSet.add(new Tag(tagName));
+        return new TagSetBuilder(tagSet);
+    }
+
+    /**
+     * Returns a {@code TagSetBuilder} with tags having tagNames appended
+     * @param tagNames to be made into {@code Tag} objects and appended
```
Comment (lightz96) [inline src/test/java/thrift/testutil/TagSetBuilder.java:69]: Do leave an empty line between description and parameter section.

**Review by lightz96 — CHANGES_REQUESTED** [review]
LGTM. I had tried running the program and could not detect any bugs. Can merge the PR once the coding standards violations are fixed.

File: src/main/java/thrift/logic/commands/TagCommand.java (hunk 1 above)
Comment (krusagiz) [reply src/main/java/thrift/logic/commands/TagCommand.java]: Fixed the indentation.

File: src/test/java/thrift/testutil/TagSetBuilder.java (hunk 2 above)
Comment (krusagiz) [reply src/test/java/thrift/testutil/TagSetBuilder.java:31]: Added the empty line.

File: src/main/java/thrift/logic/commands/UntagCommand.java (hunk 3 above)
Comment (krusagiz) [reply src/main/java/thrift/logic/commands/UntagCommand.java:52]: Added the empty line.

File: src/main/java/thrift/logic/parser/TagCommandParser.java (hunk 4 above)
Comment (krusagiz) [reply src/main/java/thrift/logic/parser/TagCommandParser.java:23]: Added the empty line.

File: src/main/java/thrift/logic/parser/UntagCommandParser.java (hunk 5 above)
Comment (krusagiz) [reply src/main/java/thrift/logic/parser/UntagCommandParser.java:23]: Added the empty line.

File: src/test/java/thrift/testutil/TagSetBuilder.java (hunk 6 above)
Comment (krusagiz) [reply src/test/java/thrift/testutil/TagSetBuilder.java:40]: Added the empty line.

File: src/test/java/thrift/testutil/TagSetBuilder.java (hunk 7 above)
Comment (krusagiz) [reply src/test/java/thrift/testutil/TagSetBuilder.java:49]: Added the empty line.

File: src/test/java/thrift/testutil/TagSetBuilder.java (hunk 8 above)
Comment (krusagiz) [reply src/test/java/thrift/testutil/TagSetBuilder.java:58]: Added the empty line.

File: src/test/java/thrift/testutil/TagSetBuilder.java (hunk 9 above)
Comment (krusagiz) [reply src/test/java/thrift/testutil/TagSetBuilder.java:69]: Added the empty line.

**Review by lightz96 — APPROVED** [review]

**Pull request merged**